### PR TITLE
Interchip comms

### PR DESCRIPTION
**Pull request by scholvat**

### CODE REVIEW

File: Autopilot/Inc/stm32f7xx_it.h [hunk 1]
```diff
@@ -56,7 +56,7 @@ void DebugMon_Handler(void);
 void SysTick_Handler(void);
 void WWDG_IRQHandler(void);
 void TIM1_UP_TIM10_IRQHandler(void);
-void TIM4_IRQHandler(void);
+void SPI1_IRQHandler(void);
```
Comment (edenfrosst) [inline Autopilot/Inc/stm32f7xx_it.h]: This will break the system. TIM4 is required. Believe this is a Cube bug.

File: Autopilot/Src/freertos.c [hunk 2]
```diff
@@ -51,7 +51,7 @@
 #include "task.h"
 #include "cmsis_os.h"
 
-/* USER CODE BEGIN Includes */
+/* USER CODE BEGIN Includes */
```
Comment (edenfrosst) [inline Autopilot/Src/freertos.c]: Unnecessary whitespace. CubeMX adds this sometimes.

File: Autopilot/Src/spi.c [hunk 3]
```diff
@@ -165,7 +165,7 @@ void HAL_SPI_MspInit(SPI_HandleTypeDef* spiHandle)
   if(spiHandle->Instance==SPI1)
   {
   /* USER CODE BEGIN SPI1_MspInit 0 */
-
+
```
Comment (edenfrosst) [inline Autopilot/Src/spi.c]: Bleh

File: Safety/Src/stm32f0xx_it.c [hunk 4]
```diff
@@ -137,6 +138,19 @@ void TIM14_IRQHandler(void)
...
+/**
+* @brief This function handles SPI1 global interrupt.
+*/
+void SPI1_IRQHandler(void)
+{
+  /* USER CODE BEGIN SPI1_IRQn 0 */
+  /* USER CODE END SPI1_IRQn 0 */
+  HAL_SPI_IRQHandler(&hspi1);
+  /* USER CODE BEGIN SPI1_IRQn 1 */
+  HAL_GPIO_WritePin(LED2_GPIO_Port, LED2_Pin, GPIO_PIN_RESET);
```
Comment (edenfrosst) [inline Safety/Src/stm32f0xx_it.c]: Required?

File: Safety/Drivers/Peripherals/Src/interchip_comm_S.c [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    dataTX = ptrTX;
+    dataRX = ptrRX;
+    //start TxRx
+    HAL_SPI_TransmitReceive_IT(&hspi1,(uint8_t *)dataTX,(uint8_t *)dataRX, sizeof(Interchip_AtoS_Packet)/sizeof(uint16_t));
+}
+
+void HAL_SPI_TxRxCpltCallback(SPI_HandleTypeDef *hspi)
+{
+  //Once packet recieved, listen for next packet
+  HAL_SPI_TransmitReceive_IT(&hspi1,(uint8_t *)dataTX,(uint8_t *)dataRX, sizeof(Interchip_AtoS_Packet)/sizeof(uint16_t));
```
Comment (edenfrosst) [inline Safety/Drivers/Peripherals/Src/interchip_comm_S.c]: Should include checks for which SPI instance is generating the callback. i.e.
`if (hspi->Instance == SPI1) {
HAL_SPI...
}`

File: Safety/Drivers/Peripherals/Src/interchip_comm_S.c [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+#include "debug.h"
+#include <stdio.h>
+#include "spi.h"
+
+
+Interchip_StoA_Packet* dataTX;
+Interchip_AtoS_Packet* dataRX;
+
+HAL_StatusTypeDef transmit_status;
+
```
Comment (edenfrosst) [inline Safety/Drivers/Peripherals/Src/interchip_comm_S.c]: Should be `static` if only used in this file.

File: Autopilot/Drivers/Peripherals/Src/interchip_comm_A.c [hunk 7]
```diff
@@ -0,0 +1,24 @@
+#include "interchip_comm_A.h"
+#include "debug.h"
+#include "spi.h"
+
+Interchip_AtoS_Packet *dataTX;
+Interchip_StoA_Packet *dataRX;
+
+HAL_StatusTypeDef transmit_status;
+
```
Comment (edenfrosst) [inline Autopilot/Drivers/Peripherals/Src/interchip_comm_A.c]: Should be `static` if only used in this file

File: Autopilot/Src/main.c [hunk 8]
```diff
@@ -129,9 +130,20 @@ int main(void)
...
+
+  Interchip_StoA_Packet* dataRX = malloc(sizeof(Interchip_StoA_Packet));
+  Interchip_AtoS_Packet* dataTX = malloc(sizeof(Interchip_AtoS_Packet));
+  interchipInit(dataTX, dataRX);
+
+  while(1){
+
+    interchipUpdate();
+    HAL_Delay(100);
+  }
```
Comment (edenfrosst) [inline Autopilot/Src/main.c]: If possible, I'd like to see this running from an OS thread.

File: Autopilot/Drivers/Peripherals/Src/interchip_comm_A.c [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    dataTX = ptrTX;
+    dataRX = ptrRX;   
+}
+
+HAL_StatusTypeDef interchipUpdate(){
+    //send data
+    HAL_StatusTypeDef transmit_status = HAL_SPI_TransmitReceive_IT(&hspi1, (uint8_t*)dataTX, (uint8_t*)dataRX, sizeof(Interchip_StoA_Packet)/sizeof(uint16_t));
+
+    if(transmit_status != HAL_OK){
+           Error_Handler();
```
Comment (edenfrosst) [inline Autopilot/Drivers/Peripherals/Src/interchip_comm_A.c]: Probably should do better error handling, but not a priority for this PR.

**Review by edenfrosst — CHANGES_REQUESTED** [review]
Some HAL usage stuff needs to be cleaned up. Probably should add comment headers to all new files (author, date, copyright, etc). Also, I'd like to know some about the reliability of this (does it crash ever, if so: how, when, how to restart it).

File: Autopilot/Src/main.c (hunk 8 above)
Comment (scholvat) [reply Autopilot/Src/main.c]: I was going to leave that for a future PR, just wanted to get the basic driver working.

File: Autopilot/Inc/stm32f7xx_it.h (hunk 1 above)
Comment (scholvat) [reply Autopilot/Inc/stm32f7xx_it.h]: Not sure what happened here... will fix

File: Safety/Drivers/Peripherals/Src/interchip_comm_S.c (hunk 5 above)
Comment (scholvat) [reply Safety/Drivers/Peripherals/Src/interchip_comm_S.c]: Yep, forgot to add that in. Shouldn't matter because there is only 1 SPI, but will fix.

File: Autopilot/Src/main.c (hunk 8 above)
Comment (edenfrosst) [reply Autopilot/Src/main.c]: Valid. Fine as is.

File: Autopilot/Drivers/Peripherals/Src/interchip_comm_A.c [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#include "interchip_comm_A.h"
+#include "debug.h"
+#include "spi.h"
+
+Interchip_AtoS_Packet *dataTX;
+Interchip_StoA_Packet *dataRX;
```
Comment (edenfrosst) [inline Autopilot/Drivers/Peripherals/Src/interchip_comm_A.c]: These should be static variables, if they're only used in this file.

File: Safety/Drivers/Peripherals/Src/interchip_comm_S.c [hunk 11]
```diff
@@ -0,0 +1,24 @@
+#include "interchip_comm_S.h"
+#include "debug.h"
+#include <stdio.h>
+#include "spi.h"
+
+
+Interchip_StoA_Packet* dataTX;
+Interchip_AtoS_Packet* dataRX;
```
Comment (edenfrosst) [inline Safety/Drivers/Peripherals/Src/interchip_comm_S.c]: These should be static variables, if they're only used in this file.

**Review by edenfrosst — CHANGES_REQUESTED** [review]
Code cleanliness issues.

**Comment by scholvat** [comment]
@ianjfrosst made things static

**Review by edenfrosst — APPROVED** [review]
